### PR TITLE
Rewrite Mesh.init_cell_orientations

### LINKED ISSUE #1038: Apparent bug with higher-order cubed-sphere
Reported by Chris Eldred

The following code for the linear shallow water equations gives a smooth solution for `meshorder = 1`, `celltype = 'quad'` or for `meshorder = 2`, `celltype = 'tri'`.  However, with the settings below, stuff goes wrong at the gridscale.

Ironically, the final result claims to be energy-preserving -- `print Kstat+Pstat`.

I'll have a go at knocking out terms, solving a mixed Poisson problem, etc etc.

```
from firedrake import *
import numpy as np

outfile = File("soln.pvd")


def make_sphere_normals():
    x = SpatialCoordinate(mesh)
    R = sqrt(inner(x, x))
    rhat_expr = x/R
    return rhat_expr
    # return CellNormal(mesh)


# settings
eltdegree = 0
meshorder = 2
rlevel = 3
celltype = 'quad'
N = 50

# constants
L = 5000. * 1000.
radius = 1.0  # 6371.22 * 1000.
g = 1.0  # 9.80616
omega = 0.5  # 7.292e-5
dt = 0.01

# create mesh
if celltype == 'quad':
    mesh = CubedSphereMesh(radius, refinement_level=rlevel, degree=meshorder)
if celltype == 'tri':
    mesh = IcosahedralSphereMesh(radius, refinement_level=rlevel, degree=meshorder)

# create operators
cell_normals = make_sphere_normals()
global_normal = Expression(("x[0]", "x[1]", "x[2]"))
mesh.init_cell_orientations(global_normal)
perp = lambda u: cross(cell_normals, u)

# create spaces
if celltype == 'quad':
    l2 = FunctionSpace(mesh, "DG", eltdegree)
    hdiv = FunctionSpace(mesh, "RTCF", eltdegree+1)
if celltype == 'tri':
    l2 = FunctionSpace(mesh, "DG", eltdegree)
    hdiv = FunctionSpace(mesh, "RT", eltdegree+1)
mixedspace = MixedFunctionSpace([l2, hdiv])

# create functions
xn = Function(mixedspace, name='xn')
xnp1 = Function(mixedspace, name='xnp1')

# set initial condition
xs = SpatialCoordinate(mesh)
x = xs[0]
y = xs[1]
z = xs[2]
H0 = 1.0  # 5960.

hinit = project(exp(-pow((-y-1.0)/0.1, 2)), l2)
uinit = project(Constant((0., 0., 0.)), hdiv)
coriolis = 2*omega*(z/radius)

hns, uns = xn.split()
hns.assign(hinit)
uns.assign(uinit)
xnp1.assign(xn)

# create forms
hn, un = split(xn)
hnp1, unp1 = TrialFunctions(mixedspace)

hstar = (hnp1 + hn)/2.
ustar = (unp1 + un)/2.
dhdt = (hnp1 - hn)/dt
dudt = (unp1 - un)/dt

hhat, uhat = TestFunctions(mixedspace)

Ms = hhat*dhdt + inner(uhat, dudt)
Sh = -hhat*H0*div(ustar)
Su = div(uhat)*g*hstar - coriolis*inner(uhat, perp(ustar))
R = (Ms - Sh - Su)*dx

M = hn*dx
K = 0.5*H0*inner(un, un)*dx
P = 0.5*g*inner(hn, hn)*dx

Mstat = np.zeros(N+1, dtype=np.float64)
Kstat = np.zeros(N+1, dtype=np.float64)
Pstat = np.zeros(N+1, dtype=np.float64)

# create solvers
a, L = system(R)
problem = LinearVariationalProblem(a, L, xnp1)
solver = LinearVariationalSolver(problem,
                                 solver_parameters={"ksp_type": "preonly",
                                                    "pc_type": "lu",
                                                    "mat_type": "aij"})

# calculate energies
Mstat[0] = assemble(M, form_compiler_parameters={'quadrature_degree': 10})
Kstat[0] = assemble(K, form_compiler_parameters={'quadrature_degree': 10})
Pstat[0] = assemble(P, form_compiler_parameters={'quadrature_degree': 10})

outfile.write(xn.sub(0), xn.sub(1))
for i in range(1, N+1):
    print i
    solver.solve()
    xn.assign(xnp1)
    outfile.write(xn.sub(0), xn.sub(1))
    Mstat[i] = assemble(M, form_compiler_parameters={'quadrature_degree': 10})
    Kstat[i] = assemble(K, form_compiler_parameters={'quadrature_degree': 10})
    Pstat[i] = assemble(P, form_compiler_parameters={'quadrature_degree': 10})
```

### PR DESCRIPTION
**Pull request by miklos1**
- `Mesh.init_cell_orientations` now works correctly for higher-order quadrilaterals. Fixes #1038.
- Test cases included, supersedes #1040.
- Replaces hand-written COFFEE AST with neat UFL.
- `Mesh.init_cell_orientations` now also accepts UFL expressions.

### CODE REVIEW

File: firedrake/mesh.py [hunk 1]
```diff
@@ -1080,8 +1081,6 @@ def init_cell_orientations(self, expr):
         import firedrake.function as function
         import firedrake.functionspace as functionspace
 
-        if expr.value_shape()[0] != 3:
-            raise NotImplementedError('Only implemented for 3-vectors')
         if self.ufl_cell() not in (ufl.Cell('triangle', 3),
                                    ufl.Cell("quadrilateral", 3),
                                    ufl.TensorProductCell(ufl.Cell('interval'), ufl.Cell('interval'),
```
Comment (wence-) [inline firedrake/mesh.py]: Does the UFL version now work for general dimension?  e.g. can you use it to call `init_cell_orientations` on a `CircleManifoldMesh`?

**Comment by wence-** [comment]
I realised, can you fix the doc example in variational-problems.rst?  Otherwise fine.

**Comment by miklos1** [comment]
@wence-: I don't see what's wrong with `variational-problems.rst`.

**Comment by wence-** [comment]
It uses the (now-deprecated) C string expression in the example code, no?

**Comment by miklos1** [comment]
@wence-: C string expressions still work with `Mesh.init_cell_orientations`. I may update `Expression` use in docs and demos later, but that does not have to happen in this pull request.

File: firedrake/mesh.py [hunk 2]
```diff
@@ -1091,59 +1090,19 @@ def init_cell_orientations(self, expr):
...
-        body.append(ast.Assign("orientation[0][0]", ast.Ternary(ast.Less("dot", 0), 1, 0)))
-
-        kernel = op2.Kernel(ast.FunDecl("void", "cell_orientations",
-                                        [ast.Decl("int**", "orientation"),
-                                         ast.Decl("double**", "coords")],
-                                        ast.Block(body)),
-                            "cell_orientations")
-
-        # Build the cell orientations as a DG0 field (so that we can
-        # pass it in for facet integrals and the like)
```
Comment (dorugeber) [inline firedrake/mesh.py]: Any reason this comment got wiped?

File: firedrake/mesh.py [hunk 3]
```diff
@@ -1091,59 +1090,19 @@ def init_cell_orientations(self, expr):
...
+        if isinstance(expr, expression.Expression):
```
Comment (dorugeber) [inline firedrake/mesh.py]: Should there be a sanity check here in the case `expr` is just a normal UFL expression?

File: firedrake/mesh.py (hunk 2 above)
Comment (wence-) [inline firedrake/mesh.py]: Although I wrote that comment, I agree that there is no need to keep it.

**Comment by wence-** [comment]
@miklos1 fine.

**Pull request merged**